rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+---
+authors: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+
+# RFD 0203 - Database Health Checks
+
+## Required Approvals
+
+- Engineering: @r0mant && @greedy52
+- Product: @roraback
+
+## What
+
+Teleport database agents will periodically perform health checks to a database's endpoint and report the results as a health status in the agent's `db_server` heartbeat.
+
+When a Teleport proxy routes a user connection, the proxy will prioritize a `healthy` agent->DB status over "less healthy" agent->DB status if multiple agents proxy the same database.
+
+## Why
+
+The primary motivation for this change is to enable smarter proxy->agent routing when multiple DB agents heartbeat the same database (i.e., an HA setup [^1]).
+
+[^1]: https://goteleport.com/docs/enroll-resources/database-access/guides/ha/#combined-replicas
+
+Currently, in an HA DB agent configuration, the proxy will randomly choose one of the DB agents to handle a user connection.
+
+If a subset of the DB agents can't reach the database endpoint, then user connection attempts will randomly fail.
+
+To fix this, the Teleport proxy service can sort the DB agents by target health status and preferentially dial agents that have healthy network connectivity to the database.
+
+Secondary motivations for this change:
+
+1. improve database resource onboarding UX
+2. surface agent->resource connectivity issues early
+3. improve troubleshooting
+4. lay the groundwork for collecting agent->resource latency measurements that can be used to make routing decisions based on latency
+
+Related issues:
+- [Database Access upstream DB health checks](https://github.com/gravitational/teleport/issues/20544)
+- [Desktop Access: support latency-based routing to a desktop service](https://github.com/gravitational/teleport/issues/40905)
+- [Detect and report latency measurements for Desktop sessions in the UI](https://github.com/gravitational/teleport/issues/35691)
+
+## Details
+
+### UX
+
+The web UI will be updated to leverage database health status information.
+
+Unhealthy databases will be displayed with a warning icon or tooltip icon indicating that the database is unhealthy.
+
+Recall that there may be multiple `db_server` heartbeats for the same database, but the web UI only shows a single resource tile for the database.
+If this is the case, then the tooltip will be shown if any of the heartbeats have an unhealthy `target_health.status`.
+The tooltip content may look something like this (or equivalent singular form verbiage if there's only one agent):
```

Comment:
The tooltip might be ok to start with but I think eventually this sort of info needs to go on the resource's own "details" page. Similar to how we have status pages for integrations. cc @roraback